File: tests/load/gcp/prom-query-grafana-dashboards.tpl.py
#!/usr/bin/env python3

from selenium import webdriver
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.common.by import By
import time
import urllib.request

from selenium.webdriver.firefox.service import Service as FirefoxService
from webdriver_manager.firefox import GeckoDriverManager
from selenium.webdriver.firefox.options import Options as FirefoxOptions

"""
Login and open a grafana dashboard using a headless browser instance.
"""


interval = ${REFRESH_INTERVAL}  # grafana refresh interval that matches prometheus scrape interval
period = 5 * 60  # reload entire dashboard every 5 minutes


def get_next_sleep_duration() -> float:
    """Calculates the remaining inter-reload sleep duration.

    This is needed in order to have all browser instances more or less synced with each other to
    query grafana server at roughly the same time.
    """
    def get_next_multiple(num, multiple):
        return num + (multiple - (num % multiple))

    tic = time.time()
    toc = get_next_multiple(tic, period)
    return toc - tic


def sleep_until_interval_elapses():
    sleep_duration = get_next_sleep_duration()
    print(f"Sleeping until interval elapses ({sleep_duration:.1f} sec)...", flush=True)
    time.sleep(sleep_duration)


def main():
    options = FirefoxOptions()
    options.headless = True
    # TODO add incognito?

    try:
        executable_path = GeckoDriverManager().install()
        # executable_path = ChromeDriverManager().install()
    except Exception as e:
        print(f"Exception '{type(e)}' raised: {str(e)}")
        time.sleep(5)
        # systemd will restart the process
        return

    GRAFANA_ADMIN_PASSWORD = urllib.request.urlopen("${COS_URL}:8081/helper/grafana/password", data=None, timeout=10.0).read().decode()

    while True:
        try:
            print("Opening up a browser instance...", flush=True)
            service = FirefoxService(executable_path)
            with webdriver.Firefox(service=service, options=options) as driver:
                driver.get("${GRAFANA_URL}/dashboards")

                # wait for login page to load
                print("Logging in...", flush=True)
                user = WebDriverWait(driver, timeout=30).until(
                    lambda d: d.find_element(By.NAME, "user"))
                password = WebDriverWait(driver, timeout=30).until(
                    lambda d: d.find_element(By.NAME, "password"))
                log_in = WebDriverWait(driver, timeout=30).until(
                    lambda d: d.find_element(By.XPATH, "//*[text()='Log in']"))

                user.send_keys("admin")
                password.send_keys(GRAFANA_ADMIN_PASSWORD)
                log_in.click()

                print("Opening dashboard...", flush=True)
                dashboard = WebDriverWait(driver, timeout=120).until(
                    lambda d: d.find_element(By.XPATH, "//*[text()='sre mock 6 panels - rates']"))
                dashboard.click()

                print("'Looking' at dashboard...", flush=True)
                time.sleep(2 * interval)

        except Exception as e:
            print(f"Exception '{type(e)}' raised: {str(e)}")
            print("Waiting for delay to elapse before next iteration", flush=True)

        sleep_until_interval_elapses()


if __name__ == "__main__":
    main()

File: tests/load/gcp/prom-query-grafana-dashboards.tpl.ts
// To run file:
// npx @flood/element-cli run example.ts --no-headless
// element run prom-query-grafana-dashboards.ts --config prom-query-grafana-dashboards.config.js

import { step, By, Until, TestSettings } from '@flood/element'

export const settings: TestSettings = {
  description: "Simulate N SREs load up a dashboard from scratch at the same time",
  stages: [
    { duration: '2m', target: 5 },
    { duration: '2m', target: 10 },
    { duration: '2m', target: 20 },
    { duration: '5d', target: 20 },
  ],
}

export default () => {
  step('1. Log in', async browser => {
    await browser.visit('${GRAFANA_URL}/dashboards')

    const user = By.nameAttr('user')
    const password = By.nameAttr('password')
    const button = By.visibleText('Log in')

    // const grafana_admin_password_response = await fetch('${COS_URL}:8081/helper/grafana/password');
    // const grafana_admin_password = await grafana_admin_password_response.text()
    
    // const grafana_admin_password = await browser.evaluate(async () => { const data = await (await fetch('http://pd-ssd-4cpu-16gb.us-central1-a.c.lma-light-load-testing.internal:8081/helper/grafana/password')).json(); return data })

    await browser.wait(Until.elementIsVisible(user))
    await browser.wait(Until.elementIsVisible(password))
    await browser.wait(Until.elementIsVisible(button))

    await browser.type(user, 'admin')
    await browser.type(password, 'GRAFANA_ADMIN_PASSWORD_TO_SUBSTITUTE_BY_CLOUD_INIT')
    await browser.click(button)
  })

  step('2. Open dashboard', async browser => {
    let dashboard = By.visibleText('sre mock 6 panels - rates')
    await browser.wait(Until.elementIsVisible(dashboard))
    await browser.click(dashboard)
  })

  step('3. Look at dashboard', async browser => {
    // await browser.wait(2 * ${REFRESH_INTERVAL})
    const entire_dashboard_reload_period = 5 * 60  // 5 minutes [sec]
    const min_delay_to_encounter_a_refresh = 2 * ${REFRESH_INTERVAL}  // [sec]
    const time_to_sleep = Math.max(min_delay_to_encounter_a_refresh, entire_dashboard_reload_period - min_delay_to_encounter_a_refresh)  // [sec]

    await new Promise(f => setTimeout(f, time_to_sleep * 1000));
  })
}
